include/sys/arc_impl.h

```diff
@@ -176,6 +176,178 @@ typedef struct l1arc_buf_hdr {
 	abd_t			*b_pabd;
 } l1arc_buf_hdr_t;
 
+typedef enum l2arc_dev_hdr_flags_t {
+	L2ARC_DEV_HDR_EVICT_FIRST = (1 << 0)	/* mirror of l2ad_first */
+} l2arc_dev_hdr_flags_t;
+
+/*
+ * Pointer used in persistent L2ARC (for pointing to log blocks & ARC buffers).
+ */
+typedef struct l2arc_log_blkptr {
+	/*
+	 * Offset of log block within the device, in bytes
+	 */
+	uint64_t	lbp_daddr;
+	/*
+	 * lbp_prop has the following format:
+	 *	* logical size (in sectors)
+	 *	* physical (compressed) size (in sectors)
+	 *	* compression algorithm (we always LZ4-compress l2arc logs)
+	 *	* checksum algorithm (used for lbp_cksum)
+	 */
+	uint64_t	lbp_prop;
+	zio_cksum_t	lbp_cksum;	/* checksum of log */
+} l2arc_log_blkptr_t;
+
+/*
+ * The persistent L2ARC device header.
+ * Byte order of magic determines whether 64-bit bswap of fields is necessary.
+ */
+typedef struct l2arc_dev_hdr_phys {
+	uint64_t	dh_magic;	/* L2ARC_DEV_HDR_MAGIC */
+	uint64_t	dh_version;	/* Persistent L2ARC version */
+	zio_cksum_t	dh_self_cksum;	/* fletcher4 of fields below */
+
+	/*
+	 * Global L2ARC device state and metadata.
+	 */
+	uint64_t	dh_spa_guid;
+	uint64_t	dh_alloc_space;		/* vdev allocated bytes */
+	uint64_t	dh_flags;		/* l2arc_dev_hdr_flags_t */
+
+	/*
+	 * Start of log block chain. [0] -> newest log, [1] -> one older (used
+	 * for initiating prefetch).
+	 */
+	l2arc_log_blkptr_t	dh_start_lbps[2];
+	uint16_t		dh_log_blk_ent;	/* entries per log blk */
+
+	const uint16_t		dh_pad[171];		/* pad to 512 bytes */
+} l2arc_dev_hdr_phys_t;
+CTASSERT_GLOBAL(sizeof (l2arc_dev_hdr_phys_t) == SPA_MINBLOCKSIZE);
+
+/*
+ * A single ARC buffer header entry in a l2arc_log_blk_phys_t.
+ */
+typedef struct l2arc_log_ent_phys {
+	dva_t			le_dva;		/* dva of buffer */
+	uint64_t		le_birth;	/* birth txg of buffer */
+	/*
+	 * le_prop has the following format:
+	 *	* logical size (in sectors)
+	 *	* physical (compressed) size (in sectors)
+	 *	* compression algorithm
+	 *	* checksum algorithm (used for le_cksum)
+	 *	* object type (used to restore arc_buf_contents_t)
+	 *	* protected status (used for encryption)
+	 *	* prefetch status (used in l2arc_read_done())
+	 */
+	uint64_t		le_prop;
+	uint64_t		le_daddr;	/* buf location on l2dev */
+	zio_cksum_t		le_cksum;	/* checksum of log entry */
+	const uint32_t		le_pad[14];	/* pad to 128 bytes	 */
+} l2arc_log_ent_phys_t;
+
+#define	L2ARC_LOG_BLK_MAX_SIZE			(128 * 1024)	/* 128k */
+#define	L2ARC_LOG_BLK_HEADER_LEN		(128)
+#define	L2ARC_LOG_BLK_MAX_ENTRIES		/* 1023 entries */	\
+	((L2ARC_LOG_BLK_MAX_SIZE - L2ARC_LOG_BLK_HEADER_LEN) /		\
+	sizeof (l2arc_log_ent_phys_t))
+
+/*
+ * A log block of up to 1023 ARC buffer log entries, chained into the
```

Comment:
Yes it would be simpler. However, this imposes a limitation of 16GB (1023 entries * SPA_MAXBLOCKSIZE) for the L2ARC to be restored. If the L2ARC device is less than 16GB it will not be restored. @behlendorf  suggested to make the entries in a log block dependent on the size of the L2ARC device, which I find very reasonable, and implemented a couple of days ago.
We determine how many entries a log block has at line 8905 of `module/zfs/arc.c`.

Reply:
@ahrens I understand what you mean now. The log blocks we write have always L2ARC_LOG_BLK_MAX_ENTRIES entries. However only some of them are actually written and restored depending on the size of L2ARC.

Reply:
So within a l2arc device, all the log blocks are the same size, which is set by dh_log_blk_ent. That makes sense.

Reply:
Yes, within a l2arc device all the log blocks are the same size, L2ARC_LOG_BLK_MAX_SIZE. How many entries are written and restored in each log block is set by `dh_log_blk_ent`.

Reply:
L2arc devices cannot change size, right?